Sources/Components/CarouselView/CarouselView.swift

```diff
@@ -0,0 +1,203 @@
+            return
+        }
+
+        collectionView.scrollToItem(at: IndexPath(item: indexPath.item + 1, section: 0), at: scrollPosition, animated: animated)
+    }
+}
+
+extension CarouselView: UICollectionViewDataSource {
+    public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        guard let numberOfItems = dataSource?.numberOfItems(in: self) else {
+            return 0
+        }
+
+        if numberOfItems > 1 {
+            return numberOfItems + 2
+        } else {
+            return numberOfItems
+        }
+    }
+
+    public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        guard let dataSource = dataSource else {
+            preconditionFailure("Data source is not available")
+        }
+
+        return dataSource.carouselView(self, cellForItemAt: self.indexPath(forItem: indexPath.item))
+    }
+}
+
+extension CarouselView: UICollectionViewDelegateFlowLayout {
+    public func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        return bounds.size
+    }
+
+    public func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        delegate?.carouselView(self, didSelectItemAt: self.indexPath(forItem: indexPath.item))
+    }
+
+    public func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
+        delegate?.carouselViewDidEndDecelerating(self)
+    }
+
+    public func scrollViewWillEndDragging(_ scrollView: UIScrollView, withVelocity velocity: CGPoint, targetContentOffset: UnsafeMutablePointer<CGPoint>) {
+        guard scrollView.isPagingEnabled else {
+            return
+        }
+
+        let numberOfItems = collectionView.numberOfItems(inSection: 0)
+        let targetPage = Int(targetContentOffset.pointee.x / scrollView.bounds.width)
+        let deltaOffset = CGFloat(numberOfItems - 2) * scrollView.bounds.width
+
+        if targetPage == 0 {
+            targetContentOffset.pointee = CGPoint(
+                x: deltaOffset,
+                y: targetContentOffset.pointee.y
+            )
+
+            scrollView.contentOffset = CGPoint(
+                x: scrollView.contentOffset.x + deltaOffset,
+                y: scrollView.contentOffset.y
+            )
+
+        } else if targetPage == numberOfItems - 1 {
+            targetContentOffset.pointee = CGPoint(
+                x: scrollView.bounds.width,
+                y: targetContentOffset.pointee.y
+            )
+
+            scrollView.contentOffset = CGPoint(
+                x: scrollView.contentOffset.x - deltaOffset,
+                y: scrollView.contentOffset.y
+            )
+        }
+    }
+}
+
+// MARK: - Private Functions
+
+private extension CarouselView {
+    func indexPath(forItem item: Int) -> IndexPath {
```

Comment:
I don't understand what this method is doing — is `item` 1-indexed? And why is the returned value called `next`?

Reply:
Oh, now I understand — it translates the Index Paths from the underlying `UICollectionView` to the logical indices used by the carousel view. Could we rename the method to better reflect this?

```
func translate(indexPath: IndexPath) -> IndexPath
func convertToCarouselIndexPath(_: IndexPath) -> IndexPath
func translateToCarousel(indexPath: IndexPath) -> IndexPath
```

Reply:
I like the `translate(_ indexPath: IndexPath) -> IndexPath`. Changing it!